### ISSUE TITLE
To header incorrectly encoded?

### ISSUE THREAD

**Issue by dmaicher** [swiftmailer/swiftmailer#650]
We stumbled across a problem when sending emails with a certain name:

```php
$message->setTo([
    'test@example.com' => 'Hotel "zum heiteren Ösi"'
]);
```

This leads to the email just being send to "Hotel". So no domain at all.

We are using the ```mail``` transport. This is the ```$to``` value actually passed to ```Swift_Transport_SimpleMailInvoker::mail```

```Hotel "zum heiteren =?utf-8?Q?=C3=96si=22?= <test@example.com>``` 

This looks weird to me?

The problem seem to be the quotes AND the german character Ö. With just the quotes or just the Ö it works fine.

Any ideas?

**Comment by dmaicher** [comment]
I did another test with smtp transport:

```php
<?php

require_once 'vendor/autoload.php';

$mailer = new Swift_Mailer(Swift_SmtpTransport::newInstance());

$message = Swift_Message::newInstance('test', 'test')
    ->setTo('test@example.com', 'Hotel "zum heiteren Ösi"')
    ->setFrom('test@example.com')
;

$mailer->send($message);
```

This leads to this being sent via port 25:

```
MAIL FROM:<test@example.com>
250 2.1.0 Ok
RCPT TO:<test@example.com>
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
Message-ID: <7d17d2be15fdd7d1abb17c08f83db1d8@swift.generated>
Date: Mon, 24 Aug 2015 20:49:27 +0200
Subject: test
From: test@example.com
To: Hotel "zum heiteren =?utf-8?Q?=C3=96si=22?= <test@example.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

test
.
250 2.0.0 Ok: queued as 5116829D5
QUIT
```

Then I did another test:

```php
$message = Swift_Message::newInstance('test', 'test')
    ->setTo('test@example.com', mb_encode_mimeheader('Hotel "zum heiteren Ösi"', 'utf-8', 'Q'))
    ->setFrom('test@example.com')
;
```

Which leads to this being sent:

```
MAIL FROM:<test@example.com>
250 2.1.0 Ok
RCPT TO:<test@example.com>
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
Message-ID: <c97628a6c0bd8c8501e7a736e45848ec@swift.generated>
Date: Mon, 24 Aug 2015 21:04:10 +0200
Subject: test
From: test@example.com
To: "Hotel \"zum heiteren =?UTF-8?Q?=C3=96si=22?=" <test@example.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

test
.
250 2.0.0 Ok: queued as A04BC29D5
QUIT
```

I'm not really familiar with RFC2822 and all the encodings but the latter output seems to be more correct to me? :)

**Comment by dmaicher** [comment]
I did another test with PHPMailer:

```php
<?php
require 'vendor/phpmailer/phpmailer/PHPMailerAutoload.php';

$mail = new PHPMailer;

$mail->isSMTP();
$mail->Host = 'localhost';
$mail->SMTPAuth = false;
$mail->SMTPSecure = '';
$mail->SMTPAutoTLS = false;
$mail->Port = 25;

$mail->CharSet = 'utf-8';
$mail->setFrom('test@example.com');
$mail->addAddress('test@example.com', 'Hotel "zum heiteren Ösi"');
$mail->Subject = 'test';
$mail->Body = 'test';
$mail->Encoding = 'quoted-printable';

$mail->send();
```

This is the result:

```
MAIL FROM:<test@example.com>
250 2.1.0 Ok
RCPT TO:<test@example.com>
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
Date: Tue, 25 Aug 2015 23:21:31 +0200
To: =?utf-8?Q?Hotel_=22zum_heiteren_=C3=96si=22?= <test@example.com>
From: test@example.com
Subject: test
Message-ID: <c2d169cb8a5afb54065762f02776e47c@ca-dev-container>
X-Priority: 3
X-Mailer: PHPMailer 5.2.10 (https://github.com/PHPMailer/PHPMailer/)
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

test
.
250 2.0.0 Ok: queued as 2CCE744E
QUIT
```

So the correct encoded To-header seems to be this one?

```=?utf-8?Q?Hotel_=22zum_heiteren_=C3=96si=22?= <test@example.com>```

**Comment by dmaicher** [comment]
Anyone?

**Comment by dmaicher** [comment]
Hello?

**Comment by adam-ca** [comment]
:+1: Hopefully they'll fix this soon!

**Comment by roomcays** [comment]
I do confirm, that using special characters and (double) quotes causes this error. I think that some MTA will have problems sending such messages, i.e. I have just faced this issue **because** of messages not being delivered.

**Comment by dmaicher** [comment]
I found another problem recently with the encoding of this sender name:

`This is „a test"`

**Comment by dmaicher** [comment]
@fabpot could I ask someone to have a look at this? We are having quite some issues in production with this :cry:

**Comment by fabpot** [comment]
Anyone can try to find a fix for this issue. Unfortunately, I won't have time to have a look at this in the near future. But as you have a solid test case, it should be possible to find the issue without too much effort.

**Comment by dmaicher** [comment]
Thank you for the fast reply @fabpot. I proposed a fix in https://github.com/swiftmailer/swiftmailer/pull/774.

**Comment by TheDigitalOrchard** [comment]
I found it surprising that SwiftMailer (even as of v5.x) was not correctly composing mail header bodies.

`From:  Joe (Black) Samson <joe@domain.com>`

This is invalid because the parentheses ( and ) have special meanings in mail headers. I would have expected SwiftMailer to correctly wrap that text in double quotes to avoid the problem.

`From: "Joe (Black) Sampson" <joe@domain.com>`

This seems to be valid. This was hard to track down because not every MTA was having a problem with this, so this went unnoticed in our software until switching MTAs.

At this point, we have to manually wrap the text in double-quotes when passing into SwiftMailer, which seems odd. Again, a library should handle this correctly. Does anyone know if the latest v6.x version of the library handles this correctly?

**Comment by roomcays** [comment]
This is weird because in the issue's fix (db057bd, e236bc4) there are already parentheses included in the regex match: `()<>\[\]:;@\,."` and it should be available since v5.4.3. Maybe the problem occurs when there are both: doublequotes and parentheses used?
